src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,910 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import {
+  getEmptyProfile,
+  getEmptyThread,
+} from '../../../profile-logic/data-structures';
+import BinaryPlistParser, { UID } from './BinaryPlistParser';
+import BinReader from './BinReader';
+import MaybeCompressedReader from './MaybeCompressedReader';
+
+//types
+import type {
+  Profile,
+  FuncTable,
+  FrameTable,
+  StackTable,
+} from '../../../types/profile';
+import type { UniqueStringArray } from '../../../../src/utils/unique-string-array';
+
+type InstrumentsSample = {|
+  +timestamp: number,
+  +threadID: number,
+  +backtraceID: number,
+  backtraceStack?: Array<number>,
+|};
+
+type TraceDirectoryTree = {|
+  +name: string,
+  +files: Map<string, File>,
+  +subdirectories: Map<string, TraceDirectoryTree>,
+|};
+
+type InstrumentsFrameInfo = {|
+  +key: string | number,
+  +name: string,
+  file?: string,
+  +line?: number,
+  +col?: number,
+|};
+
+type FormTemplateRunData = {|
+  +number: number,
+  +addressToFrameMap: Map<number, InstrumentsFrameInfo>,
+|};
+
+function readAsArrayBuffer(file: File): Promise<ArrayBuffer> {
+  return MaybeCompressedReader.fromFile(file).readAsArrayBuffer();
+}
+
+function readAsText(file: File): Promise<string> {
+  return MaybeCompressedReader.fromFile(file).readAsText();
+}
+
+function parseBinaryPlist(bytes) {
+  const text = 'bplist00';
+  for (let i = 0; i < 8; i++) {
+    if (bytes[i] !== text.charCodeAt(i)) {
+      throw new Error('File is not a binary plist');
+    }
+  }
+
+  return new BinaryPlistParser(
+    new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
+  ).parseRoot();
+}
+
+function decodeUTF8(bytes: Uint8Array): string {
+  const textDecoder = new TextDecoder();
+  if (bytes[bytes.length - 1] === 0) {
+    // Remove a single trailing null byte if present.
+    return textDecoder.decode(bytes.subarray(0, -1));
+  }
+  return textDecoder.decode(bytes);
+}
+
+function followUID(objects: any[], value: any): any {
+  return value instanceof UID ? objects[value.index] : value;
+}
+
+// This function constructs an object from given interpreter class and property list
+function patternMatchObjectiveC(
+  objects: any[],
+  value: any,
+  interpretClass: ($classname: string, obj: any) => any = x => x
```

Comment:
I think the default `interpretClass` is wrong; I feel like the second parameter should be returned?
Also maybe we don't need a default value because it's always passed?

Reply:
@julienw got it. Made the changes.